plugins/newsdownloader.koplugin/internaldownloadbackend.lua

```diff
@@ -0,0 +1,48 @@
+local http = require("socket.http")
+local https = require("ssl.https")
+local logger = require("logger")
+local ltn12 = require("ltn12")
+local socket_url = require("socket.url")
+
+local InternalDownloadBackend = {}
+local max_redirects = 10;
+
+function InternalDownloadBackend:getResponseAsString(url, redirectCount)
+    if not redirectCount then
+        redirectCount = 0
+    elseif redirectCount == max_redirects then
+        error("InternalDownloadBackend: reached max redirects: ", redirectCount)
+    end
+
+    local request, sink = {}, {}
+    request['sink'] = ltn12.sink.table(sink)
+    request['url'] = url
+    local parsed = socket_url.parse(url)
+
+    local httpRequest = parsed.scheme == 'http' and http.request or https.request
+    -- first argument returned by skip is code
+    local _, headers, status = socket.skip(1, httpRequest(request))
+
+    if status ~= "HTTP/1.1 200 OK" then
+        logger.dbg("InternalDownloadBackend: HTTP response code <> 200. Response code: ", status)
+        if status and string.sub(status, 1, 1) ~= "3" then -- handle 301, 302...
```

Comment:
Unless I'm misreading this makes no sense as written. It should be `== 3`, not not `~= 3`. But since the status code is a number you can just do `code > 300 and code < 400`. Just use it instead of ignoring it up above on line 24.

Also are you sure that this will only apply to HTTPS: that is, regular HTTP only returns something after all the internal redirections have already been taken care of?

Reply:
right copy-paste from pseudocode.

actually as status is something like:
"HTTP/1.1 200 OK"
"HTTP/1.1 400 Bad Request"
it should be :
string.sub(status, **9**, 1) **=**= "3"

Reply:
something wrong with that. I give extra debug, substring without second arg:
  logger.dbg("InternalDownloadBackend: HTTP response code <> 200. Response code: ", status)
        local code = string.sub(status, 10)
        logger.dbg("RESPO:", code)

gives:
> 01/29/18-22:09:24 DEBUG InternalDownloadBackend: HTTP response code <> 200. Response code:  HTTP/1.1 301 Moved Permanently
01/29/18-22:09:24 DEBUG RESPO: 301 Moved Permanently

adding second arg to substring:
logger.dbg("InternalDownloadBackend: HTTP response code <> 200. Response code: ", status)
        local code = string.sub(status, 10, **1**)
        logger.dbg("RESPO:", code)

gives:
> 01/29/18-22:12:09 DEBUG InternalDownloadBackend: HTTP response code <> 200. Response code:  HTTP/1.1 301 Moved Permanently
01/29/18-22:12:09 DEBUG RESPO: 
./luajit: plugins/newsdownloader.koplugin/internaldownloadbackend.lua:38: InternalDownloadBackend: Don't know how to handle HTTP status: 

see here 01/29/18-22:12:09 DEBUG RESPO: <- it's blank!!! empty null

Reply:
But you should use the number code, not the human readable string status. Up above where you're explicitly dismissing it in the dummy variable `_`, replace that with something like `code` and use it the way I said: `code > 300 and code < 400`

Reply:
But status returns me full string status:
"HTTP/1.1 200 OK"
"HTTP/1.1 400 Bad Request"

I don't have number (code only) status:( 

found ugly but working workaround.